lib/datasource/pypi/index.ts

```diff
@@ -180,21 +195,35 @@ async function getSimpleDependency(
     logger.trace({ dependency: packageName }, 'pip package not found');
     return null;
   }
-  const root: HTMLElement = parse(dep.replace(/<\/?pre>/, '')) as any;
+  const root: HTMLElement = parse(cleanSimpleHtml(dep)) as any;
   const links = root.querySelectorAll('a');
-  const releases = new Map<string, Release>();
+  const releases: Releases = {};
   for (const link of Array.from(links)) {
     const version = extractVersionFromLinkText(link.text, packageName);
     if (version) {
-      const release = releases.get(version) || { version };
-      const isDeprecated = link.hasAttribute('data-yanked');
-      if (isDeprecated) {
-        release.isDeprecated = isDeprecated;
+      const release: PypiJSONRelease = {
+        yanked: link.hasAttribute('data-yanked'),
+      };
+      const requiresPython = link.getAttribute('data-requires-python');
+      if (requiresPython) {
+        release.requires_python = requiresPython;
+      }
+      if (!releases[version]) {
+        releases[version] = [];
       }
-      releases.set(version, release);
+      releases[version].push(release);
     }
   }
-  dependency.releases = [...releases.values()];
+  const versions = compatibleVersions(releases, compatibility);
+  dependency.releases = versions.map((version) => {
```

Comment:
Is this meant as the de-duplication? How is it achieving it if you map 1:1 between inputs and outputs?

Reply:
Oh actually `versions` is already de-duplicated. The `releases` object constructed [here](https://github.com/renovatebot/renovate/pull/6966/files#diff-de5edfc564f5473b7f2a5a8c48e82629R211-R214) is a map of versions -> array of releases. After the compatibility check, this function just de-duplicates releases per compatible version and returns one release per version.

It's similar to the map function [here](https://github.com/renovatebot/renovate/pull/6966/files#diff-de5edfc564f5473b7f2a5a8c48e82629R125-R136) in `getDependency`

Reply:
If this works then let's merge it. In future I would like to move compatibility concepts out of pypi and into the generic datasource layer anyway